balancer_conn_wrappers.go

```diff
@@ -160,37 +196,119 @@ func (ccb *ccBalancerWrapper) handleSubConnStateChange(sc balancer.SubConn, s co
 	if sc == nil {
 		return
 	}
-	ccb.updateCh.Put(&scStateUpdate{
+	ccb.updateCh.Put(watcherUpdate{typ: updateTypeSubConnState, update: &scStateUpdate{
 		sc:    sc,
 		state: s,
 		err:   err,
-	})
+	}})
 }
 
-// updateClientConnState forwards the clientConn update to the wrapped balancer
-// synchronously.
-//
-// Other calls from the channel like exitIdle() and handleSubConnStateChange()
-// are handled asynchronously by pushing the update onto a channel, which is
-// picked up by the watcher() goroutine and forwarded to the wrapped balancer.
-// That approach cannot be taken here because the corresponding API on the
-// balancer returns an error which needs to be sent back to the channel to be
-// forward to the resolver.
-func (ccb *ccBalancerWrapper) updateClientConnState(ccs *balancer.ClientConnState) error {
+// handleSubConnStateChange handles a SubConnState update from the update
+// channel and invokes the appropriate method on the underlying balancer.
```

Comment:
What's the reason to not push this to the channel?
And document that.

And if we do all the outgoing calls in the watcher goroutine, maybe we don't need the balancerMu?
There were two methods that need to happen inline: Close() and ClientConnStateChange() (because of the return error). But both are handled now.

Reply:
> What's the reason to not push this to the channel?

SubConn state change is pushed onto the channel in `updateSubConnState()`. This is the method which is called when the `watcher` reads that update from the channel.

> And if we do all the outgoing calls in the watcher goroutine, maybe we don't need the balancerMu?

Woohoo !!